### PR TITLE
Send deployment event when update or rollback

### PR DESCRIPTION
**Pull request by yanganto**
- Send deployment event when update or rollback
- Add `system-info` nixosModule
- Add `system-info` command for `kld-mgr` and `kld-cli`

Same feature as https://github.com/kuutamolabs/near-staking-knd/pull/681

### CODE REVIEW

File: kld/src/cli/main.rs [hunk 1]
```diff
@@ -118,37 +126,64 @@ fn main() {
...
+        | Command::ListChannels
+        | Command::ListPeers
+        | Command::ConnectPeer { .. }
+        | Command::DisconnectPeer { .. }
+        | Command::OpenChannel { .. }
+        | Command::SetChannelFee { .. }
+        | Command::CloseChannel { .. }
+        | Command::NetworkNodes { .. }
+        | Command::NetworkChannels { .. } => {
+            if let (Some(target), Some(cert_path), Some(macaroon_path)) =
```
Comment (JosephGoulden) [inline kld/src/cli/main.rs:143]: I get that these are not required for SystemInfo command, but they will always be set in our deployments. So I think its best to keep it simple and not make them optional args.

File: kld/src/cli/main.rs [hunk 2]
```diff
@@ -118,37 +126,64 @@ fn main() {
...
+                    Command::NetworkChannels { id } => api.list_network_channels(id)?,
+                    _ => unreachable!(),
+                }
+            } else {
+                anyhow::bail!("`target`, `cert-path`, `macaroon-path` are required")
+            }
+        }
+        Command::SystemInfo { inline } => {
+            system_info(inline);
+            "null".to_string()
```
Comment (JosephGoulden) [inline kld/src/cli/main.rs:183]: Can system_info() return a string instead?

File: mgr/src/main.rs [hunk 3]
```diff
@@ -218,6 +227,29 @@ fn reboot(_args: &Args, reboot_args: &RebootArgs, config: &Config) -> Result<()>
     mgr::reboot(&hosts)
 }
 
+fn system_info(args: &SystemInfoArgs, config: &Config) -> Result<()> {
+    let hosts = filter_hosts(&args.hosts, &config.hosts)?;
+    for host in hosts {
+        let target = host.deploy_ssh_target();
+        let args = vec![target.as_str(), "--", "kld-cli", "system-info"];
+        if let Ok(output) = std::process::Command::new("ssh").args(&args).output() {
```
Comment (JosephGoulden) [inline mgr/src/main.rs:235]: Command is already imported.

File: kld/src/cli/main.rs [hunk 4]
```diff
@@ -106,6 +108,12 @@ enum Command {
         #[arg(long)]
         id: Option<String>,
     },
+    /// Show system info
+    SystemInfo {
+        /// Cli will show system info with inline format
```
Comment (JosephGoulden) [inline kld/src/cli/main.rs:113]: I suggest "Display inline format".

File: kld/src/cli/main.rs (hunk 1 above)
Comment (yanganto) [reply kld/src/cli/main.rs:143]: If that, what should we pass for `cert_path` and `macaroon_path` here

https://github.com/kuutamolabs/lightning-knd/blob/512006a1c79e4ecfac50807561e9767cea1859c3/mgr/src/nixos_rebuild.rs#L85-L87

And the code in nixos_rebuild.rs will not be simple, and we need to pass not really used parameters `cert_path` and `macaroon_path` into the function.

File: mgr/src/nixos_rebuild.rs [hunk 5]
```diff
@@ -71,5 +71,29 @@ pub fn nixos_rebuild(
...
+        "echo",
+        "level=info",
+        "$(",
+        "kld-cli",
+        "system-info",
+        "--inline",
+        ")",
+    ];
+
+    let output = std::process::Command::new("ssh").args(&args).output()?;
```
Comment (JosephGoulden) [inline mgr/src/nixos_rebuild.rs:91]: Sorry I meant in this file std::process::Command is already imported.

File: kld/src/cli/main.rs (hunk 1 above)
Comment (JosephGoulden) [reply kld/src/cli/main.rs:143]: Ok fair enough.

File: kld/src/cli/main.rs [hunk 6]
```diff
@@ -118,37 +126,64 @@ fn main() {
...
+                    Command::CloseChannel { id } => api.close_channel(id)?,
+                    Command::NetworkNodes { id } => api.list_network_nodes(id)?,
+                    Command::NetworkChannels { id } => api.list_network_channels(id)?,
+                    _ => unreachable!(),
+                }
+            } else {
+                anyhow::bail!("`target`, `cert-path`, `macaroon-path` are required")
+            }
+        }
+        Command::SystemInfo { inline } => {
```
Comment (JosephGoulden) [inline kld/src/cli/main.rs]: You can put this match arm at the top and just use underscore to match everything else.

**Comment by yanganto** [comment]
@JosephGoulden thanks for review.

**Comment by JosephGoulden** [comment]
Hi @yanganto
I deployed this to test it and it reports that kld-cli is not installed on two of the hosts, which is true. But I still want to know which version those hosts are at. So I think it would be better if kld-mgr read the toml file from each host and printed the git sha and date. 
kld-cli should only report the CARGO_PKG_VERSION.
Also kld-mgr knows which host has bitcoind deployed so it should check that directly, not through kld-cli.

**Comment by yanganto** [comment]
This ideal is like this
```
[root@validator-00:~]# kneard-ctl system-info
kneard-version: 0.3.0
git-sha: 5cf3faaeab0faa33d3298c4e8d5b9a472ca7b8b4
git-commit-date: 20230425112147
neard-version: nix:1.33.0-rc.1
neard-protocol-version: 60
neard-db-version: 35
```

We can use `*-mgr` to get the version without `*-cli`, but it is not true.  Because it always gets the latest commit of a branch, not the real system running on the node.  

The current problem is why ` kld-cli`  is not installed.  Did I modify something that make `kld-cli` fail to install?

If we want to know the git-sha from a `.toml` but not the real system running on the node, we can check out Github, I think we do not really need that feature.

**Comment by JosephGoulden** [comment]
Kld is only installed on one node because it is not designed for high availability, that work is still to be done.
I was thinking mgr should still read from the toml file on the node - that is updated with each deployment.
I eventually wanted to see the git revision in the API, so that would also read the toml file, but the logic for that will be in the core of the node. So having the cli read the toml file will seem odd then I think.

**Comment by yanganto** [comment]
Hmm, there will be at least one `kld` node, we can ask `system-info` from this kld node.  I still think keeping the mgr of near and of btc the same behavior is a good thing. So there will no error message to complain on `kld-cli` absent when `kld-mgr system-info`.

The corresponding git hash of the toml is already in the `flake.lock`.  We do not need a redundant `system-info` for it, we can just read `flake.lock`.  

When there are main and dev, if any issue happens in main, the first step to address the problem is to know the versions running on the node, and this is what `system-info` aim for. 

Is this good for you?

**Comment by paulmattei** [comment]
Does it make sense to introduce a 'kld-ctl'?

**Comment by yanganto** [comment]
The `kld-cli` is like `kneard-ctl`, but it is kld-node only.

Here is the architecture, and now `kld-mgr` will send different command dependence on the node type.

```txt
For near
`kneard-mgr`(user's machine) -(ssh)->  `kneard-ctl`(remote node) -(socket)-> `kneard`(remode node)

For Btc
`kld-mgr`(user's machine) -(ssh)->  `kld-cli`(remote kld node) -(socket)-> `kld`(remode kld node)
`kld-mgr`(user's machine) -(ssh)->  `cockroach-sql`/`cockroach-rpc` -(socket)-> `cockroachdb`(remode db node)
```

**Comment by paulmattei** [comment]
Not really though because -ctl is not user facing, whereas -cli is...I wondered if that's what's causing the confusion...

**Comment by yanganto** [comment]
Okay, I see. The command will move into `kld-ctl`.  And this tool will be installed in every node.

File: nix/modules/kld/default.nix [hunk 7]
```diff
@@ -157,6 +158,10 @@ in
   config = {
     # for cli
     environment.systemPackages = [ kld-cli bitcoin-cli ];
+    environment.etc."system-info.toml".text = lib.mkDefault ''
+      git_sha = "${self.rev or "dirty"}"
+      git_commit_date = "${self.lastModifiedDate}"
+    '';
```
Comment (Mic92) [inline nix/modules/kld/default.nix:164]: You can simply use an activation script instead using SSH from rust.
This is less code and also works if kld-ctl breaks during upgrade.

```suggestion
    '';
    system.activationScripts.nixos-upgrade = ''
      ${config.systemd.package}/bin/systemd-run --collect --unit nixos-upgrade echo level=info message="git_sha=${} git_commit_date=${self.lastModifiedDate}"
  '';

```

**Comment by JosephGoulden** [comment]
I don't know what kld-ctl would be for or how it would help in this scenario.
I want to keep system-info. Its just that I will add this to the API so the logic of reading the toml file will be duplicated. I don't really care where kld-mgr reads the version from, flake.lock is fine.
I still think kld-mgr should call bitcoind --version directly.

**Comment by yanganto** [comment]
> I still think kld-mgr should call bitcoind --version directly.

No, it is not a good idea to call `bitcoind` directly.


Consider two architectures,  more network issues less robust for a command

If `kld-mgr` call bitcoind directly, there will be two ssh connection.

```text
`kld-mgr`  --(ssh)-> kld-cli --> /etc/system-info 
           \-(ssh)-> bitcoind
```

If kld-cli collects bitcoind version, there is only one ssh connection

```text
`kld-mgr`  --(ssh)-> kld-cli --> /etc/system-info 
                             \-> bitcoind
```

File: nix/modules/kld/default.nix (hunk 7 above)
Comment (yanganto) [reply nix/modules/kld/default.nix:164]: `CARGO_PKG_VERSION` could not be included, so the chainge is not applied.

Comment (yanganto) [reply nix/modules/kld/default.nix:164]: `CARGO_PKG_VERSION` and bitcoind version are not included for kld-node, and DB version is not included for db-node. 

We can provide these versions in the update event, once we update nodes, we can exactly know what we updated.

File: mgr/src/nixos_rebuild.rs [hunk 8]
```diff
@@ -75,5 +75,42 @@ pub fn nixos_rebuild(
...
+        ],
+    };
+
+    let output = Command::new("ssh").args(&args).output()?;
+    if !output.status.success() {
+        warn!(
+            "Fail to send deployment event: {}",
+            std::str::from_utf8(&output.stdout).unwrap_or("stdout utf-8 decode error")
+        );
+    }
```
[the remark this answers is not in the document]
Comment (yanganto) [reply mgr/src/nixos_rebuild.rs:114]: There is no `kld-ctl` in every node, we need to match on node type.

If we do not want the match code here, we still need some node type-dependent code in another place.

File: nix/modules/kld/default.nix (hunk 7 above)
Comment (yanganto) [reply nix/modules/kld/default.nix:164]: Thank you very much, I get the idea.

**Comment by JosephGoulden** [comment]
I think you should be able to join all the commands together so that they are executed in one SSH session. By making the args something like [target.as_str(), "--", "cat", "/etc/system-info.toml", "&&", "kld-cli", "system-info", "&&", "bitcoind", "--version", "&&", "cockroach-sql", "&&", "--version"].

Toml dependency is probably not required as we are just reading a very simple file.

Also the kld node also has a cockroach instance installed so that should be printed out as well.

**Comment by yanganto** [comment]
Semi-structured data is already done.  `system-info` command also helps information clear and ready for integration.  I take near testnet for example, because these are machines I can reach.

- `kneard-mgr system-info --hosts=validator-00`  shows information for humans (user), this command also allows showing more hosts at the same time.
```text
[validator-00]
kneard-version: 0.3.0
git-sha: 230aeb64b387bb45e1f40d5e6ba8b4316f466ab5
git-commit-date: 20230503055722
neard-version: nix:1.33.0-rc.1
neard-protocol-version: 60
neard-db-version: 35
```

- `kneard-ctl system-info`  shows information for a human(a logined developer)
```
kneard-version: 0.3.0
git-sha: 230aeb64b387bb45e1f40d5e6ba8b4316f466ab5
git-commit-date: 20230503055722
neard-version: nix:1.33.0-rc.1
neard-protocol-version: 60
neard-db-version: 35

```
- `kneard-ctl system-info --inline` shows information for other cli tool to integrate
```text
kneard-version=0.3.0 git-sha=230aeb64b387bb45e1f40d5e6ba8b4316f466ab5 git-commit-date=20230503055722 neard-version=nix:1.33.0-rc.1 neard-protocol-version=60 neard-db-version=35
```

What I think we eventually need but not implement now
`kneard-ctl system-info --json` shows information for end user in some dashboard.


We can do the way without any digesting.
` [target.as_str(), "--", "cat", "/etc/system-info.toml", "&&", "kld-cli", "system-info", "&&", "bitcoind", "--version", "&&", "cockroach-sql", "&&", "--version"]`

```text
git-sha = "5cf3faaeab0faa33d3298c4e8d5b9a472ca7b8b4"
git-commit-date = "20230425112147"
Bitcoin Core version v24.0.1
Copyright (C) 2009-2022 The Bitcoin Core developers

Please contribute if you find Bitcoin Core useful. Visit
<https://bitcoincore.org/> for further information about the software.
The source code is available from <https://github.com/bitcoin/bitcoin>.

This is experimental software.
Distributed under the MIT software license, see the accompanying file COPYING
or <https://opensource.org/licenses/MIT>
Bitcoin Core version v24.0.1
Copyright (C) 2009-2022 The Bitcoin Core developers

Please contribute if you find Bitcoin Core useful. Visit
<https://bitcoincore.org/> for further information about the software.
The source code is available from <https://github.com/bitcoin/bitcoin>.

This is experimental software.
Distributed under the MIT software license, see the accompanying file COPYING
or <https://opensource.org/licenses/MIT>
```

This information makes people a headache and easily be ignored, could we consider structurize the information on the fly?

I will take care cockroach db on kld-node.

**Comment by JosephGoulden** [comment]
The API returns structured JSON for to be consumed by other apps including a dashboard where the version will be shown.

For the manager I would just use some simple bash commands to get the data into the desired format.
`cockroach-sql version | head -1 | sed 's/Build Tag: */Cockroach SQL version: /' && bitcoind --version | head -1 | sed 's/version/version:/'`

We don't want to keep modifying kld-cli when we add and remove other services. KLD is just one service like cockroach or bitcoind. The manager is the thing that knows which services are deployed where. So whether it be with bash commands or something more sophisticated it should know how to find the versions of the components it has deployed.

**Comment by Mic92** [comment]
I don't have a strong opinion whether it should there should be a system-info subcommand or simply cat is used. However I would like to keep things consistent between here and https://github.com/kuutamolabs/near-staking-knd/pull/681

**Comment by JosephGoulden** [comment]
Regarding the naming. I would just suggest that at least the kld-cli system-info to be renamed to version. As that is all that is being returned, and 'version' is more usual and everyone expects programs to have this command.

Bitcoind-cli and kld-cli already have a get-info command which returns more general information about the service, including the version, but do require API authentication.

**Comment by yanganto** [comment]
> I would just suggest that at least the kld-cli system-info to be renamed to version. As that is all that is being returned, and 'version' is more usual and everyone expects programs to have this command.

It is sorry to say no here because of the following point.

- `version` did not have much usual sense for people, there are `--version`, `-V`, which you can try on the command line tools on GNU/Linux.

- `kld-cli`, every `*-cli`, `*-mgr` already have `--version`, we do not need a redundant `version` one or a different output `version` command to confuse people

- It already includes `hash`, `LastModifyDate` not only versions now, and possibly more different things in the future.

- the way to return the version is much more related to language and framework. But the `version` is not fit for us.

Rust and Nix ecosystem is using `--version` more, and Go is using `version` but it isn't generally used on our repos of all organizations.  It does not make sense to change any language default to make every program provide the same command. The version format is also different.  We should not take an effort to make a different language act the same, if this mater, we can choose Rust implementation, it is better memory control than Go implementation.


> Bitcoind-cli and kld-cli already have a get-info command which returns more general information about the service, including the version, but do require API authentication.

If you suggest `system-info` should be covered by `get-info`, we can move these into `get-info`.  In a user view, he will wonder why needs an authentication when reading versions, hash, and last modified date. So it is still better to keep `system-info` a different one from `get-info`.

File: kld/src/cli/main.rs (hunk 1 above)
Comment (JosephGoulden) [reply kld/src/cli/main.rs:132]: Actually these should remain required args. The manager is calling kld-cli via SSH so it is using the wrapper script that sets the cert_path and macaroon_path to the readonly macaroon.

**Comment by JosephGoulden** [comment]
Ok I forgot --version is generated automatically. So seeing as that exists, and reports the package version I think that can be called by the manager.
Hash, etc can be read from the system-info.toml by the manager as discussed.
Yeah I think we should remove system-info from kld-cli.
The version stuff can be moved to get-info as part of the API call. I can do that part if you like, up to you.
I don't think any users will be wondering about authentication, it will already be set up for them by the nix wrapper script.

File: kld/src/cli/main.rs (hunk 1 above)
Comment (yanganto) [reply kld/src/cli/main.rs:132]: Are `cert_path` and `macaroon_path` meaningful?
As I known, `cert_path` is `/var/lib/kld/certs/ca.pem`, and `macaroon_path` is `/var/lib/kld/macaroons/admin.macaroon`.  These just are paths, event if the content change the path are still the same.

**Comment by yanganto** [comment]
Hi @JosephGoulden, Do `kld-ctl` looks good now?
After a full discussion, I think `kld-ctl` is the best way to do this, and we will not touch `kld-cli`.

**Comment by JosephGoulden** [comment]
I think another binary - kld-ctl seems a bit overkill at this point, but I prefer that to including the logic in kld-cli.

kld-ctl would not be "controlling" anything - its just aggregating the versions from the different services. And controller is a synonym for manager so the naming is confusing.

I would still just do this with bash either from kld-mgr or a create a version script on the server. But if you want to make kld-ctl to be consistent with near then I can live with that.

**Review by JosephGoulden — APPROVED** [review]

File: ctl/src/system_info.rs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+struct SystemInfo {
+    git_sha: String,
+    git_commit_date: String,
+}
+
+fn read_system_info() -> Result<SystemInfo> {
+    if let Ok(content) = std::fs::read_to_string("/etc/system-info.toml") {
+        Ok(toml::from_str::<SystemInfo>(&content)?)
+    } else {
+        bail!("fail to read /etc/system-info.toml")
```
Comment (Mic92) [inline ctl/src/system_info.rs:15]: Use .context here to not throw away the underlying error.

File: ctl/src/system_info.rs [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+        .output()?;
+    if output.status.success() {
+        if let Some(version) = std::str::from_utf8(&output.stdout)?
+            .split('\n')
+            .next()
+            .and_then(|line| line.split("Bitcoin Core version ").nth(1))
+        {
+            Ok(version.into())
+        } else {
+            bail!("fail get version from return bitcoind")
```
Comment (Mic92) [inline ctl/src/system_info.rs:31]: .context

File: ctl/src/system_info.rs [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+        }
+    } else {
+        bail!("fail to get version from bitcoind")
```
Comment (Mic92) [inline ctl/src/system_info.rs:34]: .context

File: ctl/src/system_info.rs [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+        .output()?;
+    if output.status.success() {
+        if let Some(version) = std::str::from_utf8(&output.stdout)?
+            .split('\n')
+            .next()
+            .and_then(|line| line.split("Build Tag:        ").nth(1))
+        {
+            Ok(version.into())
+        } else {
+            bail!("fail get version from return bitcoind")
```
Comment (Mic92) [inline ctl/src/system_info.rs:50]: .context

File: ctl/src/system_info.rs [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+        }
+    } else {
+        bail!("fail to get version from bitcoind")
```
Comment (Mic92) [inline ctl/src/system_info.rs:53]: .context

File: ctl/src/system_info.rs [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+fn bitcoind_version() -> Result<String> {
+    let output = std::process::Command::new("bitcoind")
+        .args(["--version"])
+        .output()?;
+    std::str::from_utf8(&output.stdout)?
+        .split('\n')
+        .next()
+        .and_then(|line| line.split("Bitcoin Core version ").nth(1))
+        .ok_or(anyhow!("fail get version from return bitcoind"))
```
Comment (Mic92) [inline ctl/src/system_info.rs:25]:
```suggestion
        .ok_or(anyhow!("failed to parse version from bitcoind command output"))
```

File: ctl/src/system_info.rs [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+fn cockroach_version() -> Result<String> {
+    let output = std::process::Command::new("cockroach")
+        .args(["version"])
+        .output()?;
+    std::str::from_utf8(&output.stdout)?
+        .split('\n')
+        .next()
+        .and_then(|line| line.split("Build Tag:        ").nth(1))
+        .ok_or(anyhow!("fail get version from return cockroach"))
```
Comment (Mic92) [inline ctl/src/system_info.rs:37]:
```suggestion
        .ok_or(anyhow!("failed to parse version from cockroach output"))
```

File: ctl/src/system_info.rs [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+fn read_system_info() -> Result<SystemInfo> {
+    let content = std::fs::read_to_string("/etc/system-info.toml")
+        .context("fail to read /etc/system-info.toml")?;
+    Ok(toml::from_str::<SystemInfo>(&content)?)
+}
+
+fn bitcoind_version() -> Result<String> {
+    let output = std::process::Command::new("bitcoind")
+        .args(["--version"])
+        .output()?;
```
Comment (Mic92) [inline ctl/src/system_info.rs:20]:
```suggestion
        .output().context("could not run bitcoind command")?;
```

File: ctl/src/system_info.rs [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+        .next()
+        .and_then(|line| line.split("Bitcoin Core version ").nth(1))
+        .ok_or(anyhow!("fail get version from return bitcoind"))
+        .map(|version| version.into())
+}
+
+fn cockroach_version() -> Result<String> {
+    let output = std::process::Command::new("cockroach")
+        .args(["version"])
+        .output()?;
```
Comment (Mic92) [inline ctl/src/system_info.rs:32]:
```suggestion
        .output().context("could not run cockroach command");
```

**Comment by yanganto** [comment]
bors merge

**Pull request merged**